# Exploratory Data Analysis With SQLite

## INTRODUCE DATABASE
In this project, we'll work with data from the [CIA World Factbook](https://www.cia.gov/the-world-factbook/), a compendium of statistics about all of the countries on Earth. The Factbook contains demographic information like the following:
- **population:** the global population.
- **population_growth:** the annual population growth rate, as a percentage.
- **area:** the total land and water area.

[Download database here](https://dsserver-prod-resources-1.s3.amazonaws.com/257/factbook.db)

## 1.Load Database

In [1]:
%%capture
%load_ext sql
%sql sqlite:///factbook.db

'Connected: None@factbook.db'

In [2]:
%%sql
SELECT *
  FROM sqlite_master
 WHERE type='table';

Done.


[('table', 'sqlite_sequence', 'sqlite_sequence', 3, 'CREATE TABLE sqlite_sequence(name,seq)'),
 ('table', 'facts', 'facts', 47, 'CREATE TABLE "facts" ("id" INTEGER PRIMARY KEY AUTOINCREMENT NOT NULL, "code" varchar(255) NOT NULL, "name" varchar(255) NOT NULL, "area" integer, "a ... (4 characters truncated) ... land" integer, "area_water" integer, "population" integer, "population_growth" float, "birth_rate" float, "death_rate" float, "migration_rate" float)')]

## 2.Overview of The Data

In [3]:
%%sql
SELECT *
FROM facts
LIMIT 5;

Done.


[(1, 'af', 'Afghanistan', 652230, 652230, 0, 32564342, 2.32, 38.57, 13.89, 1.51),
 (2, 'al', 'Albania', 28748, 27398, 1350, 3029278, 0.3, 12.92, 6.58, 3.3),
 (3, 'ag', 'Algeria', 2381741, 2381741, 0, 39542166, 1.84, 23.67, 4.31, 0.92),
 (4, 'an', 'Andorra', 468, 468, 0, 85580, 0.12, 8.13, 6.96, 0.0),
 (5, 'ao', 'Angola', 1246700, 1246700, 0, 19625353, 2.78, 38.78, 11.49, 0.46)]

- **name**: the name of the country.
- **area**: the country's total area (both land and water).
- **area_land**: the country's land area in square kilometers.
- **area_water**: the country's waterarea in square kilometers.
- **population**: the country's population.
- **population_growth**: the country's population growth as a percentage.
- **birth_rate**: the country's birth rate, or the number of - births per year per 1,000 people.
- **death_rate**: the country's death rate, or the number of death per year per 1,000 people.

## 3.Summary Statistics

### 3.1.Which country has the highest population? And Population Growth ?

In [4]:
%%sql
SELECT MIN(population) AS "Min Population",
        MAX(population) AS "Max Population",
        MIN(population_growth) AS "Min Population Growth",
        MAX(population_growth) AS "Max Population Growth"
FROM facts        

Done.


[(0, 7256490011, 0.0, 4.02)]

A few things stick out from the summary statistics::
 - There is a country population ~ 0
 - There is a country population ~ 7.2 bilion

In [5]:
%%sql
SELECT name, population, population_growth
FROM facts
WHERE population == (SELECT MAX(population) FROM facts)
        OR population_growth == (SELECT MAX(population_growth) FROM facts)
        OR population == (SELECT MIN(population) FROM facts)

Done.


[('South Sudan', 12042910, 4.02),
 ('Antarctica', 0, None),
 ('World', 7256490011, 1.08)]

There are two countries that have special values:
- **Antarctica:** According to Wikipedia, Antarctica has no permanent inhabitants. It seems like the table contains a row for Antarctica, which explains the population of 0
- **World:** the row contains the whole world, which explains the maximum population of over 7.2 billion we found earlier

> ***We should calculate the summary statistics we calculated earlier while excluding the row for the whole world.***

In [6]:
%%sql
SELECT  CASE
        WHEN population == (SELECT MAX(population) 
                                   FROM facts
                                   WHERE name != 'World') THEN "The Highest Population: " || name
        WHEN population_growth == (SELECT MAX(population_growth) 
                                   FROM facts
                                   WHERE name != 'World') THEN "The Highest Population Growth: "|| name 
        ELSE NULL
        END AS 'COUNTRY', population, population_growth
FROM facts
ORDER BY COUNTRY DESC
LIMIT 2

Done.


[('The Highest Population: China', 1367485388, 0.45),
 ('The Highest Population Growth: South Sudan', 12042910, 4.02)]

- **South Sudan** has the highest population growth
- **China** has the highest population

### 3.2.Which countries have the highest ratios of water to land? 

In [7]:
%%sql
SELECT  name , area_water, area_land, MAX(ROUND(CAST(area_water AS FLOAT)/area_land,2)) AS 'ratios_water_to_land'
FROM facts

Done.


[('British Indian Ocean Territory', 54340, 60, 905.67)]

**British Indian Ocean Territory** has the highest ratios of water to land

### 3.3.Which countries have more water than land?

In [8]:
%%sql
SELECT name, area_water, area_land
FROM facts
WHERE area_water > area_land 
    OR (area_water NOT NULL AND area_land IS NULL)
ORDER BY area_water DESC

Done.


[('Ethiopia', 104300, None),
 ('British Indian Ocean Territory', 54340, 60),
 ('Virgin Islands', 1564, 346)]

There are three countries that have more water than land are **Ethiopia**, **British Indian Ocean Territory**, and **Virgin Islands**.

### 3.4.Which countries will add the most people to their populations next year?

In [9]:
%%sql
SELECT name, migration_rate 
FROM facts
WHERE migration_rate == (SELECT MAX(migration_rate)
                         FROM facts
                         WHERE name != 'World')

Done.


[('Qatar', 22.39)]

With migration rate ~ 22.39 (highest), **Qatar** will add the most people to their populations next year  

### 3.5.Which countries have a higher death rate than birth rate?

In [10]:
%%sql
SELECT name, death_rate,  birth_rate
FROM facts
WHERE death_rate > birth_rate AND name != 'World'
ORDER BY death_rate

Done.


[('Monaco', 9.24, 6.65),
 ('Austria', 9.42, 9.41),
 ('Japan', 9.51, 7.93),
 ('Saint Pierre and Miquelon', 9.72, 7.42),
 ('Bosnia and Herzegovina', 9.75, 8.87),
 ('Italy', 10.19, 8.74),
 ('Poland', 10.19, 9.74),
 ('Czech Republic', 10.34, 9.63),
 ('Portugal', 11.02, 9.27),
 ('Greece', 11.09, 8.66),
 ('Slovenia', 11.37, 8.42),
 ('Germany', 11.42, 8.47),
 ('Romania', 11.9, 9.14),
 ('Croatia', 12.18, 9.45),
 ('Estonia', 12.4, 10.51),
 ('Moldova', 12.59, 12.0),
 ('Hungary', 12.73, 9.16),
 ('Belarus', 13.36, 10.7),
 ('Serbia', 13.66, 9.08),
 ('Russia', 13.69, 11.6),
 ('Lithuania', 14.27, 10.1),
 ('Latvia', 14.31, 10.0),
 ('Bulgaria', 14.44, 8.92),
 ('Ukraine', 14.46, 10.72)]


## 4.Finding Densely Populated Countries

### 4.1.The Densely Populated Countries have:
- **Above** average values for **population**.
- **Below** average values for **area**.

In [11]:
%%sql
SELECT ROUND(AVG(population)) AS 'Average Population', ROUND(AVG(area)) AS 'Average Area'
FROM facts
WHERE name != 'World'

Done.


[(32242667.0, 555094.0)]

In [12]:
%%sql
SELECT name, population, area
FROM facts
WHERE population > (SELECT AVG(population)
                   FROM facts
                   WHERE name != 'World')
AND  area < (SELECT AVG(area)
                   FROM facts
                   WHERE name != 'World')

Done.


[('Bangladesh', 168957745, 148460),
 ('Germany', 80854408, 357022),
 ('Iraq', 37056169, 438317),
 ('Italy', 61855120, 301340),
 ('Japan', 126919659, 377915),
 ('Korea, South', 49115196, 99720),
 ('Morocco', 33322699, 446550),
 ('Philippines', 100998376, 300000),
 ('Poland', 38562189, 312685),
 ('Spain', 48146134, 505370),
 ('Thailand', 67976405, 513120),
 ('Uganda', 37101745, 241038),
 ('United Kingdom', 64088222, 243610),
 ('Vietnam', 94348835, 331210)]

### 4.2.Which countries have the highest population/area ratio?

In [13]:
%%sql
SELECT name, population, area, ROUND(CAST(population AS FLOAT)/area,2) AS Population_Density
FROM facts
WHERE area NOT NULL AND population NOT NULL
        AND area != 0 AND population != 0
ORDER BY Population_Density DESC;

Done.


[('Macau', 592731, 28, 21168.96),
 ('Monaco', 30535, 2, 15267.5),
 ('Singapore', 5674472, 697, 8141.28),
 ('Hong Kong', 7141106, 1108, 6445.04),
 ('Gaza Strip', 1869055, 360, 5191.82),
 ('Gibraltar', 29258, 6, 4876.33),
 ('Bahrain', 1346613, 760, 1771.86),
 ('Maldives', 393253, 298, 1319.64),
 ('Malta', 413965, 316, 1310.02),
 ('Bermuda', 70196, 54, 1299.93),
 ('Sint Maarten', 39689, 34, 1167.32),
 ('Bangladesh', 168957745, 148460, 1138.07),
 ('Guernsey', 66080, 78, 847.18),
 ('Jersey', 97294, 116, 838.74),
 ('Barbados', 290604, 430, 675.82),
 ('Mauritius', 1339827, 2040, 656.78),
 ('Taiwan', 23415126, 35980, 650.78),
 ('Aruba', 112162, 180, 623.12),
 ('Lebanon', 6184701, 10400, 594.68),
 ('Saint Martin', 31754, 54, 588.04),
 ('San Marino', 33020, 61, 541.31),
 ('Korea, South', 49115196, 99720, 492.53),
 ('Rwanda', 12661733, 26338, 480.74),
 ('West Bank', 2785366, 5860, 475.32),
 ('Nauru', 9540, 21, 454.29),
 ('Tuvalu', 10869, 26, 418.04),
 ('Netherlands', 16947904, 41543, 407.96),
 ('Marshall Islands', 72191, 181, 398.85),
 ('Israel', 8049314, 20770, 387.55),
 ('Burundi', 10742276, 27830, 386.0),
 ('India', 1251695584, 3287263, 380.77),
 ('Belgium', 11323973, 30528, 370.94),
 ('Haiti', 10110019, 27750, 364.33),
 ('Comoros', 780971, 2235, 349.43),
 ('Philippines', 100998376, 300000, 336.66),
 ('Sri Lanka', 22053488, 65610, 336.13),
 ('Japan', 126919659, 377915, 335.84),
 ('Curacao', 148406, 444, 334.25),
 ('Grenada', 110694, 344, 321.78),
 ('Guam', 161785, 544, 297.4),
 ('El Salvador', 6141350, 21041, 291.88),
 ('Vietnam', 94348835, 331210, 284.86),
 ('American Samoa', 54343, 199, 273.08),
 ('Jamaica', 2950210, 10991, 268.42),
 ('Saint Lucia', 163922, 616, 266.11),
 ('Saint Vincent and the Grenadines', 102627, 389, 263.82),
 ('United Kingdom', 64088222, 243610, 263.08),
 ('Puerto Rico', 3598357, 13791, 260.92),
 ('Pakistan', 199085847, 796095, 250.08),
 ('Trinidad and Tobago', 1222363, 5128, 238.37),
 ('Liechtenstein', 37624, 160, 235.15),
 ('Germany', 80854408, 357022, 226.47),
 ('British Virgin Islands', 33454, 151, 221.55),
 ('Luxembourg', 570252, 2586, 220.52),
 ('Dominican Republic', 10478756, 48670, 215.3),
 ('Nepal', 31551305, 147181, 214.37),
 ('Cayman Islands', 56092, 264, 212.47),
 ('Antigua and Barbuda', 92436, 442, 209.13),
 ('Korea, North', 24983205, 120538, 207.26),
 ('Italy', 61855120, 301340, 205.27),
 ('Seychelles', 92430, 455, 203.14),
 ('Sao Tome and Principe', 194006, 964, 201.25),
 ('Saint Kitts and Nevis', 51936, 261, 198.99),
 ('Switzerland', 8121830, 41277, 196.76),
 ('Nigeria', 181562056, 923768, 196.55),
 ('Qatar', 2194817, 11586, 189.44),
 ('Andorra', 85580, 468, 182.86),
 ('Anguilla', 16418, 91, 180.42),
 ('Gambia, The', 1967709, 11300, 174.13),
 ('Kosovo', 1870981, 10887, 171.85),
 ('Kuwait', 2788534, 17818, 156.5),
 ('Uganda', 37101745, 241038, 153.92),
 ('Isle of Man', 87545, 572, 153.05),
 ('Malawi', 17964697, 118484, 151.62),
 ('Micronesia, Federated States of', 105216, 702, 149.88),
 ('Tonga', 106501, 747, 142.57),
 ('China', 1367485388, 9596960, 142.49),
 ('Guatemala', 14918999, 108889, 137.01),
 ('Cabo Verde', 545993, 4033, 135.38),
 ('Czech Republic', 10644842, 78867, 134.97),
 ('Indonesia', 255993674, 1904569, 134.41),
 ('Togo', 7552318, 56785, 133.0),
 ('Thailand', 67976405, 513120, 132.48),
 ('Kiribati', 105711, 811, 130.35),
 ('Denmark', 5581503, 43094, 129.52),
 ('Cyprus', 1189197, 9251, 128.55),
 ('Akrotiri', 15700, 123, 127.64),
 ('Poland', 38562189, 312685, 123.33),
 ('Dhekelia', 15700, 130, 120.77),
 ('European Union', 513949445, 4324782, 118.84),
 ('Portugal', 10825309, 92090, 117.55),
 ('Azerbaijan', 9780780, 86600, 112.94),
 ('Northern Mariana Islands', 52344, 464, 112.81),
 ('Tokelau', 1337, 12, 111.42),
 ('Slovakia', 5445027, 49035, 111.04),
 ('Ghana', 26327649, 238533, 110.37),
 ('Wallis and Futuna', 15613, 142, 109.95),
 ('Hungary', 9897541, 93028, 106.39),
 ('Albania', 3029278, 28748, 105.37),
 ('Moldova', 3546847, 33851, 104.78),
 ('France', 66553766, 643801, 103.38),
 ('A

**THERE IS A DIFFERENCE BETWEEN 4.1 AND 4.2**

## REFERENCES
- [Dataquest.io](https://github.com/dataquestio/solutions/blob/master/Mission257Solutions.ipynb)